packages/fuel-indexer-schema/src/db/graphql.rs

```diff
@@ -261,43 +261,136 @@ impl Operation {
+                    // to the queue, then check the entity of the selection against the
+                    // current nesting level. If they differ, then the operation has moved
+                    // out of a child entity into a parent entity.
+                    if let Some(current_nesting_level) = nested_entity_stack.last() {
+                        if entities.len() < last_seen_entities_len
+                            && current_nesting_level != &entity_name
+                        {
+                            let _ = nested_entity_stack.pop();
+                            elements.push(QueryElement::ObjectClosingBoundary);
                         }
-                    })
-                    .collect();
+                    }
 
-                let column_text = columns.join(", ");
+                    last_seen_entities_len = entities.len();
 
-                let namespace = format!("{namespace}_{identifier}");
-                let mut query = format!(
-                    "SELECT {} FROM {}.{}",
-                    column_text,
-                    namespace,
-                    name.to_lowercase()
-                );
+                    if let Selection::Field(field_name, _f, subselections) = current {
+                        if subselections.selections.is_empty() {
+                            elements.push(QueryElement::Field {
+                                key: field_name.clone(),
+                                value: format!(
+                                    "{namespace}_{identifier}.{entity_name}.{field_name}"
+                                ),
+                            });
+                        } else {
+                            // If the current entity has a foreign key on the current
+                            // selection, join the foreign table on that primary key
+                            // and set the field as the innermost entity by pushing to the stack.
+                            if let Some(field_to_foreign_key) =
+                                schema.foreign_keys.get(&entity_name.to_lowercase())
+                            {
+                                if let Some(foreign_key) =
+                                    field_to_foreign_key.get(&field_name.to_lowercase())
+                                {
+                                    let reference_table =
+                                        format!("{namespace}_{identifier}.{field_name}");
+                                    let referencing_key =
+                                        format!("{namespace}_{identifier}.{entity_name}.{field_name}");
+                                    let primary_key =
+                                        format!("{namespace}_{identifier}.{field_name}.{foreign_key}");
+                                    let join = format!(
+                                        "INNER JOIN {reference_table} ON {referencing_key} = {primary_key}");
```

Comment:
Similar comment here. We're doing a lot of raw-string processing - which is fine, but I think there's justification to abstract this a bit